internal/dump/common.go

```diff
@@ -75,12 +76,19 @@ func dumpTree(ctx context.Context, repo restic.Repository, rootNode *restic.Node
 func GetNodeData(ctx context.Context, output io.Writer, repo restic.Repository, node *restic.Node) error {
 	var (
 		buf []byte
+		ok  bool
 		err error
 	)
+
+	blobCache := blobcache.New(64 << 20)
 	for _, id := range node.Content {
-		buf, err = repo.LoadBlob(ctx, restic.DataBlob, id, buf)
-		if err != nil {
-			return err
+		buf, ok = blobCache.Get(id)
+		if !ok {
+			buf, err = repo.LoadBlob(ctx, restic.DataBlob, id, buf)
```

Comment:
This is broken. LoadBlob gets buf as its last argument so it can reuse the memory (that's why it's declared outside the loop). With this patch, it can reuse a buffer that comes from the cache, so the cached blobs can get overwritten.

The easy fix is to drop the separate declarations, except for err, and pass nil to LoadBlob:

```go
func GetNodeData(ctx context.Context, output io.Writer, repo restic.Repository, node *restic.Node) (err error) {
	blobCache := blobcache.New(64 << 20)
	for _, id := range node.Content {
		buf, ok := blobCache.Get(id)
		if !ok {
			buf, err = repo.LoadBlob(ctx, restic.DataBlob, id, nil)
// as before
```

That may hurt performance in the common case of no repeated blobs, though.